gcloud-java-contrib/gcloud-java-nio/src/main/java/com/google/gcloud/storage/contrib/nio/CloudStorageFileSystemProvider.java

```diff
@@ -537,8 +539,32 @@ public void createDirectory(Path dir, FileAttribute<?>... attrs) {
    */
   @Override
   public DirectoryStream<Path> newDirectoryStream(Path dir, Filter<? super Path> filter) {
-    // TODO: Implement me.
-    throw new UnsupportedOperationException();
+    CloudStoragePath cloudPath = checkPath(dir);
+    checkNotNull(filter);
+    String prefix = cloudPath.toString();
+    Iterator<Blob> iterator = storage.list(cloudPath.bucket(), Storage.BlobListOption.prefix(prefix)).iterateAll();
+    final List<Path> ret = new ArrayList<>();
+    while (iterator.hasNext()) {
+      Path p = (Path)(cloudPath.getFileSystem().getPath(iterator.next().name()));
+      try {
+        if (filter.accept(p)) {
+          ret.add(p);
+        }
+      } catch (IOException boom) {
+        // The doc says we should throw this in the iterator, but since we're prefilling
+        // the list this is the next best thing.
+        throw new DirectoryIteratorException(boom);
```

Comment:
To preserve the behavior suggested in the doc and to avoid fetching all data in advance we could make `DirectoryStream` return a lazy iterator over `storage.list(...).iterateAll()`:
Something like (just throwing an idea I did not test this):
```java
private static class LazyPathIterator extends AbstractIterator<Path> {

  private Iterator<Blob> blobIterator;
  private Filter<? super Path> filter;

  LazyPathIterator(Iterator<Blob> blobIterator, Filter<? super Path> filter) {
    this.blobIterator = blobIterator;
    this.filter = filter;
  }

  @Override
  protected Path computeNext() {
    while (blobIterator.hasNext()) {
      Path path = (Path) cloudPath.getFileSystem().getPath(iterator.next().name());
      try {
        if (filter.accept(path)) {
          return path;
        }
      } catch (IOException ex) {
        throw new DirectoryIteratorException(ex);
      }
    }
    return endOfData();
  }
}
```